mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -236,122 +522,503 @@ impl TendermintCoin {
         tx_payload: Any,
         fee: Fee,
         timeout_height: u64,
+        memo: String,
     ) -> cosmrs::Result<Raw> {
         let signkey = SigningKey::from_bytes(&self.priv_key)?;
-        let tx_body = tx::Body::new(vec![tx_payload], TX_DEFAULT_MEMO, timeout_height as u32);
+        let tx_body = tx::Body::new(vec![tx_payload], memo, timeout_height as u32);
         let auth_info = SignerInfo::single_direct(Some(signkey.public_key()), account_info.sequence).auth_info(fee);
         let sign_doc = SignDoc::new(&tx_body, &auth_info, &self.chain_id, account_info.account_number)?;
         sign_doc.sign(&signkey)
     }
+
+    pub fn add_activated_token_info(&self, ticker: String, decimals: u8, denom: Denom) {
+        self.tokens_info
+            .lock()
+            .insert(ticker, ActivatedTokenInfo { decimals, denom });
+    }
+
+    fn estimate_blocks_from_duration(&self, duration: u64) -> i64 {
+        let estimated_time_lock = (duration / self.avg_block_time as u64) as i64;
+
+        if estimated_time_lock > MAX_TIME_LOCK {
+            MAX_TIME_LOCK
+        } else if estimated_time_lock < MIN_TIME_LOCK {
+            MIN_TIME_LOCK
+        } else {
+            estimated_time_lock
+        }
+    }
+
+    pub(crate) fn check_if_my_payment_sent_for_denom(
+        &self,
+        decimals: u8,
+        denom: Denom,
+        other_pub: &[u8],
+        secret_hash: &[u8],
+        amount: &BigDecimal,
+    ) -> Box<dyn Future<Item = Option<TransactionEnum>, Error = String> + Send> {
+        let amount = try_fus!(sat_from_big_decimal(amount, decimals));
+        let amount = vec![Coin {
+            denom,
+            amount: amount.into(),
+        }];
+
+        let pubkey_hash = dhash160(other_pub);
+        let to_address = try_fus!(AccountId::new(&self.account_prefix, pubkey_hash.as_slice()));
+
+        let htlc_id = self.calculate_htlc_id(&self.account_id, &to_address, amount, secret_hash);
+
+        let coin = self.clone();
+        let fut = async move {
+            let htlc_response = try_s!(coin.query_htlc(htlc_id.clone()).await);
+            let htlc_data = match htlc_response.htlc {
+                Some(htlc) => htlc,
+                None => return Ok(None),
+            };
+
+            match htlc_data.state {
+                0 => {},
+                unexpected_state => return Err(format!("Unexpected state for HTLC {}", unexpected_state)),
+            };
+
+            let rpc_client = try_s!(coin.rpc_client().await);
+            let q = format!("create_htlc.id = '{}'", htlc_id);
+
+            let response = try_s!(
+                // Search single tx
+                rpc_client
+                    .perform(TxSearchRequest::new(q, false, 1, 1, TendermintResultOrder::Ascending))
+                    .await
+            );
+
+            if let Some(tx) = response.txs.first() {
+                if let cosmrs::tendermint::abci::Code::Err(_err_code) = tx.tx_result.code {
```

Comment:
> If we find a tx that finished with unsuccessful code, it means that payment was sent, but failed for some reason. As we return `Ok(None)`, MM2 will attempt to send it again. Is it possible that we will just burn a bit more fees for one more failed transaction in this case?
> 

I believe if it fails, then we get an error on the broadcasting attempt. Am I wrong?

> Btw, if we try to create HTLC with id that was already used, will `create_htlc.id = '{}'` query still return only one transaction? I have not noticed this during our DM conversation slightly_smiling_face

HTLCs cant have already used `htlc_id`. It will return an error on broadcasting the tx.

For example:
* [tx with id 92D17CD6BC9F32FE056F78C5490598A00B58253E618871EBD3D719A72DC9539B](https://nyancat.iobscan.io/#/tx?txHash=5650BA897E1219705F6F9447953B1E071F0B0720769A338E7AA7E3B968CCDC9D)
* [second attemp with same id](https://nyancat.iobscan.io/#/tx?txHash=EB246FC89952DD59ACF0C611442E3B704A80BE2755F965031E26988374EF2FEE)

Reply:
![htlc2022-10-21_18-51](https://user-images.githubusercontent.com/39852038/197237426-27a175e0-543c-4c7c-acc9-fe6279e0e7ac.png)

Like in the picture above, `fn send_raw_tx_bytes` returned an error and stop the process for broadcasting existing htlc_id on `test_htlc_create_and_claim` test. But it's actually available here https://nyancat.iobscan.io/#/tx?txHash=57EE62B2DF7E311C98C24AE2A53EB0FF2C16D289CECE0826CA1FF1108C91B3F9

I updated it to return an error instead of returning Ok(None) to be sure anyway.